airflow-core/src/airflow/models/dagrun.py

```diff
@@ -351,7 +352,7 @@ def __init__(
         if queued_at is NOTSET:
             self.queued_at = timezone.utcnow() if state == DagRunState.QUEUED else None
         elif queued_at is not None:
-            self.queued_at = queued_at
+            self.queued_at = cast("datetime", queued_at)
```

Comment:
I'm thinking how to avoid this `cast`. Perhaps
```python
        if queued_at is NOTSET:
            self.queued_at = timezone.utcnow() if state == DagRunState.QUEUED else None
        elif isinstance(queued_at, datetime) or queued_at is None:
            self.queued_at = queued_at
        else:
            raise TypeError
 ```

Reply:
Yeah it’s currently a limitation in Python’s type system. I’m working on some improvements as a part of another PR of mine (with a custom TypeGuard function), but for now let’s just use isinstance.

Reply:
Forgot to mention, my improvement to this has been merged, please use the `is_arg_set` helper.